### PR TITLE
feat: Web Locks API

### PR DESCRIPTION
**Pull request by load1n9**

### CODE REVIEW

**Comment by sgwilym** [comment]
Seems like this is failing on a formatting issue?

File: ext/weblocks/00_weblocks.js [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    LOCKED: 0,
+    UNLOCKED: 1,
+  };
+
+  class Lock {
+    queue = [];
+    owner = false;
+    trying = false;
+    name;
+    mode;
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:31]: these should be using symbols and getters for the ones that need them

File: ext/weblocks/00_weblocks.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    constructor(name, mode = LockMode[0], buffer = null) {
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:32]: should be a illegal constructor and use `webidl.createBranded` and assign that way.

File: ext/weblocks/00_weblocks.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      this.name = name;
+      this.mode = mode;
+      this.buffer = buffer ?? new SharedArrayBuffer(4);
+      this.flag = new Int32Array(this.buffer, 0, 1);
+      if (!buffer) {
+        Atomics.store(this.flag, 0, LockState.UNLOCKED);
+      }
+    }
+
+    enter(handler) {
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:42]: should use symbol. needs webidl brand check.

File: ext/weblocks/00_weblocks.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      return new Promise((resolve) => {
+        this.queue.push({ handler, resolve });
+        this.trying = true;
+        setTimeout(() => {
+          this.tryEnter();
+        }, 0);
+      });
+    }
+
+    tryEnter() {
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:52]: should use symbol. needs webidl brand check.

File: ext/weblocks/00_weblocks.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      this.owner = true;
+      this.trying = false;
+      const { handler, resolve } = this.queue.shift();
+      handler(this).finally(() => {
+        this.leave();
+        resolve();
+      });
+    }
+
+    leave() {
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:65]: should use symbol. needs webidl brand check.

File: ext/weblocks/00_weblocks.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+      if (!this.owner) return;
+      Atomics.store(this.flag, 0, LockState.UNLOCKED);
+      this.owner = false;
+      const message = { webLocks: true, kind: "leave", name: this.name };
+      InternalLockManager.send(message);
+      this.tryEnter();
+    }
+  }
+
+  class LockManagerSnapshot {
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:75]: not a class, but a plain object.

File: ext/weblocks/00_weblocks.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+              lock.tryEnter();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  class LockManager {
+    [webidl.brand] = webidl.brand;
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:169]: brand not necessary because of createBranded call.

File: ext/weblocks/00_weblocks.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      return InternalLockManager.send(message);
+    }
+
+    /**
+     * @param {*} message
+     */
+    receive(message) {
+      webidl.assertBranded(this, LockManagerPrototype);
+      return InternalLockManager.receive(message);
+    }
```
Comment (crowlKats) [inline ext/weblocks/00_weblocks.js:215]: I dont see these in the spec.

**Review by crowlKats — CHANGES_REQUESTED** [review]

**Comment by load1n9** [comment]
sadly I have come to believe that adding web locks now would be a terrible idea.  Writing a full weblocks implementation would require a whole system to store data across instances which eventually should be added to deno but adding it just for weblocks seems unnecessary so I believe it would be better to implement some more multithreading features before rewriting the weblocks api

**Pull request closed without merging**